File: backend/SalesApp.Api/Services/OrderMath.cs
using SalesApp.Api.Models;

namespace SalesApp.Api.Services;

public static class OrderMath
{
    /// <summary>
    /// Recomputes PaidAmount/RemainingAmount/PaymentStatus on an order from its payments.
    /// The order's Payments collection must be loaded.
    /// </summary>
    public static void Recalculate(Order order)
    {
        order.PaidAmount = order.Payments.Sum(p => p.Amount);
        // Never a negative due: any overpayment is tracked as the customer's advance balance,
        // not as a negative remaining on the order.
        order.RemainingAmount = Math.Max(0, order.TotalAmount - order.PaidAmount);

        if (order.PaidAmount <= 0)
            order.PaymentStatus = PaymentStatus.Pending;
        else if (order.PaidAmount >= order.TotalAmount)
            order.PaymentStatus = PaymentStatus.Paid;
        else if (order.Status == OrderStatus.Pending)
            order.PaymentStatus = PaymentStatus.Advance;   // partial, paid before dispatch
        else
            order.PaymentStatus = PaymentStatus.Partial;   // partial, but already dispatched/delivered
    }

    /// <summary>Amount paid against a single order, never more than the order total.</summary>
    public static decimal AppliedToOrder(Order order) => Math.Min(order.PaidAmount, order.TotalAmount);

    /// <summary>Overpayment on a single order — the part that becomes advance.</summary>
    public static decimal OverpaidOn(Order order) => Math.Max(0, order.PaidAmount - order.TotalAmount);

    /// <summary>Customer advance = money paid beyond what each order required, summed across orders.</summary>
    public static decimal AdvanceBalance(IEnumerable<Order> orders) => orders.Sum(OverpaidOn);
}

/// <summary>Reserved payment methods used by the advance-balance machinery.</summary>
public static class PaymentMethods
{
    public const string Advance = "Advance";                 // advance applied TO an order (positive)
    public const string AdvanceTransfer = "AdvanceTransfer"; // overpayment moved OUT of an order (negative)
}

File: android-app/app/src/main/java/com/salesapp/mobile/data/repo/OrderMath.kt
package com.salesapp.mobile.data.repo

import com.salesapp.mobile.data.models.OrderStatus
import com.salesapp.mobile.data.models.PaymentStatus
import java.sql.Connection
import kotlin.math.max

/**
 * Recomputes an order's PaidAmount / RemainingAmount / PaymentStatus from its Payments ledger
 * and persists them (mirrors the backend OrderMath.Recalculate).
 */
object OrderMath {

    const val METHOD_ADVANCE = "Advance"
    const val METHOD_ADVANCE_TRANSFER = "AdvanceTransfer"

    fun recalculate(conn: Connection, orderId: Int) {
        val (total, statusInt) = conn.prepareStatement("SELECT TotalAmount, Status FROM Orders WHERE Id = ?").use { ps ->
            ps.setInt(1, orderId)
            ps.executeQuery().use { if (!it.next()) return else it.getDouble(1) to it.getInt(2) }
        }
        val paid = conn.prepareStatement("SELECT COALESCE(SUM(Amount), 0) FROM Payments WHERE OrderId = ?").use { ps ->
            ps.setInt(1, orderId); ps.executeQuery().use { if (it.next()) it.getDouble(1) else 0.0 }
        }
        val remaining = max(0.0, Sql.round2(total - paid))
        val status = OrderStatus.from(statusInt)
        val paymentStatus = when {
            paid <= 0.0 -> PaymentStatus.Pending
            paid >= total -> PaymentStatus.Paid
            status == OrderStatus.Pending -> PaymentStatus.Advance
            else -> PaymentStatus.Partial
        }
        conn.prepareStatement("UPDATE Orders SET PaidAmount = ?, RemainingAmount = ?, PaymentStatus = ? WHERE Id = ?").use { ps ->
            ps.setDouble(1, Sql.round2(paid)); ps.setDouble(2, remaining); ps.setInt(3, paymentStatus.value); ps.setInt(4, orderId)
            ps.executeUpdate()
        }
    }

    /** Cancelled → all paid becomes advance; else paid over the total. */
    fun overpaidOn(status: OrderStatus, total: Double, paid: Double): Double =
        if (status == OrderStatus.Cancelled) max(0.0, paid) else max(0.0, paid - total)
}
